src/main/scala/is/hail/io/bgen/BgenRecord.scala

```diff
@@ -0,0 +1,176 @@
+    val nRow = (bytes(0) & 0xff) | ((bytes(1) & 0xff) << 8) | ((bytes(2) & 0xff) << 16) | ((bytes(3) & 0xff) << 24)
+    assert(nRow == nSamples, "row nSamples is not equal to header nSamples")
+
+    val nAlleles = (bytes(4) & 0xff) | ((bytes(5) & 0xff) << 8)
+    assert(nAlleles == expectedNumAlleles, s"Value for `nAlleles' in genotype probability data storage is not equal to value in variant identifying data. Expected $expectedNumAlleles but found $nAlleles.")
+
+    val minPloidy = bytes(6) & 0xff
+    val maxPloidy = bytes(7) & 0xff
+
+    if (minPloidy != 2 || maxPloidy != 2)
+      fatal(s"Hail only supports diploid genotypes. Found min ploidy equals `$minPloidy' and max ploidy equals `$maxPloidy'.")
+
+    val samplePloidy = Array.ofDim[Int](nSamples)
+    var i = 0
+    while (i < nSamples) {
+      val b = bytes(i + 8)
+      val isMissing = ((b & 0xffL) >>> 7) == 1
+      val ploidy = if (isMissing) -1 else b & 0x3f
+      assert(ploidy >= -1 && ploidy <= 63, s"Ploidy value must be in range [0, 63]. Found $ploidy.")
+      samplePloidy(i) = ploidy
+      i += 1
+    }
+    assert(i == nSamples, s"Number of ploidy values `$i' does not equal the number of samples `$nSamples'.")
+
+    val phase = bytes(nSamples + 8) & 0xff
+    assert(phase == 0 || phase == 1, s"Value for phase must be 0 or 1. Found $phase.")
+    val isPhased = phase == 1
+
+    if (isPhased)
+      fatal("Hail does not support phased genotypes.")
+
+    val nBitsPerProb = bytes(nSamples + 9) & 0xff
+    assert(nBitsPerProb >= 1 && nBitsPerProb <= 32, s"Value for nBits must be between 1 and 32 inclusive. Found $nBitsPerProb.")
+
+    val probabilityIterator = new BgenProbabilityIterator(bytes, nBitsPerProb, nSamples + 10)
+
+    val nGenotypes = triangle(nAlleles)
+    val sampleProbs = Array.ofDim[Long](nGenotypes)
+    val totalProbInt = ~0L >>> (64 - nBitsPerProb)
+    val noCall: Genotype = new DosageGenotype(-1, null)
+
+    val nExpectedBytesProbs = math.ceil((nSamples * (nGenotypes - 1) * nBitsPerProb).toDouble / 8).toInt
+    assert(bytes.length == nExpectedBytesProbs + nSamples + 10, s"Number of uncompressed bytes `${ bytes.length }' does not match the expected size `$nExpectedBytesProbs'.")
+
+    new Iterable[Genotype] {
+      def iterator = new Iterator[Genotype] {
+        var sampleIndex = 0
+
+        def hasNext: Boolean = sampleIndex < nSamples
+
+        def next(): Genotype = {
+          var i = 0
+          var sumProbInt = 0L
+          while (i < nGenotypes - 1) {
+            assert(probabilityIterator.hasNext, "Did not decode bytes correctly. Ran out of probabilities.")
+            val p = probabilityIterator.next()
+            sampleProbs(i) = p
+            i += 1
+            sumProbInt += p
+          }
+
+          assert(sumProbInt <= totalProbInt, "Sum of probabilities is greater than 1.")
+          sampleProbs(i) = totalProbInt - sumProbInt
+
+          val gt = if (samplePloidy(sampleIndex) == -1)
```

Comment:
No -- ploidy can be equal to -1 (I chose this as the missing value -- now that I think about it, I can just test if the ploidy is negative because missing is defined as the most significant bit set to 1). They still emit probabilities for No Calls by having two probabilities equal to 0 in the diploid/biallelic case. If the ploidy wasn't set to missing, you'd interpret the probabilities as Array(0.0, 0.0, 1.0) because the third probability is deduced from the others.